### PR TITLE
Add some more comprehensive R-Group Decomposition tests

### PR DESCRIPTION
**Pull request by greglandrum**
The R-Group Decomposition (RGD) code is complex, has a lot of options, and isn't particularly well tested. There are a number of open PRs suggesting changes to the RGD code and we really don't have a decent way to detect whether or not they are breaking things. The complexity here is probably inevitable, but I would like to try and fix the lack of testing.

I think there are two parts of this:
1. Make sure that we have unit tests in place for all possible values of all of the options (not all possible combinations... that would be nuts). Ideally these would include testing some combinations which someone who knows the code would class as edge cases.
2. Add some more complex, regression-like tests that try and reproduce things seen in real world datasets. Ideally these tests are human understandable so someone can look at the molecules, look at the tests, and figure out whether or not the correct result is being produced.

This PR is a proposal for how we might start the second part. It includes a new, catch2-based, set of tests where I've tried to make the expected results comprehensible. Here's an example
```
        CHECK(flatten_whitespace(toJSON(rows)) ==
          flatten_whitespace(readReferenceData(testDataDir + "simple1.out1.json")));
```
where `simple1.out1.json` contains:
```
[
    {
        "Core": "Cc1cccc([*:2])c1[*:1]",
        "R1": "CO[*:1]",
        "R2": "[H][*:2]"
    },
    {
        "Core": "Cc1cccc([*:2])c1[*:1]",
        "R1": "CO[*:1]",
        "R2": "[H][*:2]"
    },
    {
        "Core": "Cc1cccc([*:2])c1[*:1]",
        "R1": "[H][*:1]",
        "R2": "CO[*:2]"
    }
]
```
We're using catch2, so if there's a mismatch when the tests run, both what is observed and what was expected will be output to the console (also nice).

@bp-kelley suggested that we also have tests based on the jupyter notebook since it can be easier to see what's going on there. To that end I added a jupyter notebook - `test_data/rgd_testing.ipynb` - which can be used together with either your favorite notebook test runner (will test that things run and that the output generated matches what is in the stored JSON files) or with nbval (will tests that things run and also does doctest-style testing of the notebooks). To make that run automatically I added a new cmake function `add_jupytertest` which can be used like this:
```
add_jupytertest(jupyterRGroupDecomposition
         ${CMAKE_CURRENT_SOURCE_DIR}/../test_data rgd_testing.ipynb)
```
These use nbval and are disabled automatically if it's not installed.

I also added a jupyter notebook in the test_data directory which allows you to browse the files which are there and look at the compounds. 

I think we need more realistic test cases than just what I've added here, ideally chosen to exercise specific bits of code - like stereochemistry handling, symmetrization, etc. - but I want to check to see what you guys think of the testing format and then we can all start thinking about the tests to add.

Ideally in parallel we can also start working on the lack of unit tests.

### CODE REVIEW

**Comment by greglandrum** [comment]
@jones-gareth : it would be great if you could take a look at this too.

**Comment by bp-kelley** [comment]
I’ve been doing my tests in notebooks, do you mind if I add a notebook runner to the tests?  The visual results help me a lot for debugging.

**Comment by greglandrum** [comment]
> I’ve been doing my tests in notebooks, do you mind if I add a notebook runner to the tests? The visual results help me a lot for debugging.

It'd be nice if we also had everything available as C++ whenever possible (since those tests get run more often).
But I don't think there's a problem duplicating tests.
How are you planning on doing automated testing with notebooks? (Asking that last one because we don't want to add too many non-stable dependencies to the CI process)

**Comment by bp-kelley** [comment]
The technique I've been doing is:

```
jupyter nbconvert --execute --inplace ./test-001.ipynb --ExecutePreprocessor.timeout=18000
if [ $? -ne 0 ]
then
  echo "Failed running notebook"
  exit 1
fi
```

But I'll see if there is something better for cmake.  I would like to actually run all notebooks like this, at least occasionally to ensure they still work.

**Comment by bp-kelley** [comment]
Looks like it is trivial to make it into a unit test

https://www.blog.pythonlibrary.org/2018/10/16/testing-jupyter-notebooks/

**Comment by gedeck** [comment]
Here is code that collects the errors:
```
import nbformat
from nbconvert.preprocessors import ExecutePreprocessor
from nbconvert.preprocessors.execute import CellExecutionError
 
 
def run_notebook(notebook_path, output_path):
    print(f'Running: {notebook_path.stem:60s} ', end='', flush=True)
 
    with open(notebook_path, encoding='utf-8') as f:
        nb = nbformat.read(f, as_version=4)
         
    proc = ExecutePreprocessor(timeout=600, kernel_name='python3')
    proc.allow_errors = True
    proc.allow_errors = False

    try:
      proc.preprocess(nb, {'metadata': {'path': notebook_path.parent}})
    except CellExecutionError:
      with open(output_path / f'error_{notebook_path.stem}.ipynb', mode='wt', encoding='utf-8') as f:
          nbformat.write(nb, f)
      print('ERROR')
      return nb, []
 
    with open(output_path / f'{notebook_path.stem}.ipynb', mode='wt', encoding='utf-8') as f:
        nbformat.write(nb, f)
 
    errors = []
    for cell in nb.cells:
        cell_type = cell['cell_type']
        if cell_type in ['markdown', 'code']:
            continue
        print(cell_type)
        if cell_type == 'code':
            print(cell['outputs'])
        else:
            print(cell['cell_type'])
            print(list(cell))
            if 'outputs' not in cell: continue
            for output in cell['outputs']:
                errors.append(output)
#             if output.output_type == 'error':
#                 errors.append(output)
    print('ERROR' if errors else 'OK')
    return nb, errors
```
You will need to adjust it.

**Comment by greglandrum** [comment]
I think nbval (https://nbval.readthedocs.io/en/latest/) looks cleaner than those options

**Comment by gedeck** [comment]
nbval: that's nice, as it does the output comparison. My code was only checking that the calculations don't fail. This will be much better for my use cases too.

**Comment by greglandrum** [comment]
yeah, I've used nbval a couple of times and it's not terrible. I like that you can use tags (or comments) to specify which cells are ignored and which get tested.

**Comment by bp-kelley** [comment]
Does nbval work with unitest?  Looks like it is a py.test extension

**Comment by greglandrum** [comment]
It is a py.test extension.

**Comment by jones-gareth** [comment]
I like the JSON specs for rgroup tests.  Would it be possible to create a library of specs in a external file, so that C++, Python and notebooks can share the specs?  Often the same test systems are shared between C++ and Python and correcting errors in expected values is repetitive.

This would also make it easy to include a notebook visualization for every C++ test (perhaps as an enhancement to `View datasets.ipynb`), which would be really cool to have.

**Comment by greglandrum** [comment]
> I like the JSON specs for rgroup tests. Would it be possible to create a library of specs in a external file, so that C++, Python and notebooks can share the specs? Often the same test systems are shared between C++ and Python and correcting errors in expected values is repetitive.
> 
> This would also make it easy to include a notebook visualization for every C++ test (perhaps as an enhancement to `View datasets.ipynb`), which would be really cool to have.

Nice idea. I need to tweak what I've done so that it produces actual JSON, but that should be easy.
Will give it a try

**Comment by greglandrum** [comment]
@jones-gareth I've switched to using actual JSON, added saved .json files, and some notebook tests.

The PR description has been updated with more info about this.

File: Code/GraphMol/RGroupDecomposition/RGroupUtils.cpp [hunk 1]
```diff
@@ -103,5 +105,53 @@ bool hasDummy(const RWMol &core) {
...
+  }
+  if (!hasQuery) {
+    return MolToSmiles(mol);
+  } else {
+    return MolToSmarts(mol);
+  }
+}
 }  // namespace
 
+std::string toJSON(const RGroupRow &rgr, const std::string &prefix) {
```
Comment (bp-kelley) [inline Code/GraphMol/RGroupDecomposition/RGroupUtils.cpp:134]: We should probably add RGroupCols as well I think.

Comment (bp-kelley) [inline Code/GraphMol/RGroupDecomposition/RGroupUtils.cpp:134]: prefix isn't documented or used as far as I can tell, what is the purposes of this var?

**Review by bp-kelley — CHANGES_REQUESTED** [review]

**Comment by bp-kelley** [comment]
Good stuff, once this is merged I'll add some more tests.

**Comment by jones-gareth** [comment]
> 
> 
> @jones-gareth I've switched to using actual JSON, added saved .json files, and some notebook tests.
> 
> The PR description has been updated with more info about this.

Looks good.

File: Code/GraphMol/RGroupDecomposition/RGroupUtils.cpp (hunk 1 above)
Comment (greglandrum) [reply Code/GraphMol/RGroupDecomposition/RGroupUtils.cpp:134]: I had been thinking about this primarily as a tool for human-readable tests, not unit testing, and I think the row-based format is way better for that.
But if we're looking at it as generic functionality for converting this stuff to JSON (which it clearly is) then having the column version completely makes sense.
On it.

Comment (greglandrum) [reply Code/GraphMol/RGroupDecomposition/RGroupUtils.cpp:134]: Will document it. It's there to make things more readable... it's a string which is added to each output line.
Currently it's being used to allow indentation

**Comment by greglandrum** [comment]
@bp-kelley the changes you were looking for are in

**Review by bp-kelley — APPROVED** [review]

**Comment by greglandrum** [comment]
@ptosco do you have time to take a look at this or shall I just merge since Brian and Gareth have both given a thumbs up?

**Comment by ptosco** [comment]
@greglandrum happy for you to merge, and many thanks for the nice additions! This will be very useful going forward, really appreciated.
